include/pybind11/detail/smart_holder_type_casters.h

```diff
@@ -292,7 +292,8 @@ class modified_type_caster_generic_load_impl {
             loaded_v_h = value_and_holder();
             return true;
         }
-        if (convert && cpptype && try_as_void_ptr_capsule(src)) {
+        const auto &bases = all_type_info(srctype);
+        if (convert && cpptype && bases.empty() && try_as_void_ptr_capsule(src)) {
```

Comment:
The (possibly expensive) `all_type_info` call is needed only if `convert && cpptype` is true:
```
if (convert && cpptype) {
  const auto &bases = all_type_info(srctype);
  if (bases.empty() && try_as_void_ptr_capsule(src)) {
    return true;
  }
}
```
But even that is still doing too much work.
After everything is working (CI is green), but before marking this PR as ready for review, I'd look into something like `has_type_info(src_type)` which returns immediately when it find the first one.
Not sure if that actually makes sense, but I'd definitely look to find out.
Another idea: have a const pointer for bases right under the `Case 2` comment, if set already inside `Case 2`, reuse here.
But that one may be a completely inconsequential optimization, although also an easy one, not sure if it's worth it or not.

Reply:
There is `get_type_info`: https://github.com/pybind/pybind11/blob/master/include/pybind11/detail/type_caster_base.h#L184. But `all_type_info` is called inside. I did not find anything else that is related to this right now. But I will spend more time on this direction. 

For `all_type_info`, looks like it will cache the result: https://github.com/pybind/pybind11/blob/master/include/pybind11/detail/type_caster_base.h#L167. So I believe if it reached `Case 2`, the cached result will be returned.